File: pset6/credit/credit.py
from cs50 import get_int
# Luhn's Algorithm
#
# Multiply every other digit by 2, starting with the number’s second-to-last digit,
# and then add those products’ digits together.
# Add the sum to the sum of the digits that weren’t multiplied by 2.
# If the total’s last digit is 0 (or, put more formally, if the total modulo 10 is congruent to 0),
# the number is valid!


def main():
    # Get card number from user
    card_num = get_int("Number: ")
    total_sum = 0
    card_length = 0

    # loop until we process all card numbers
    while True:
        # Add to count the total number of digits
        card_length += 1

        # Get current number to process - truncating
        curr_num = int(card_num % 10)

        # Process as per Luhns Algorithm
        tmp = 0
        if card_length % 2 == 0:
            curr_num = curr_num * 2
            if curr_num >= 10:
                # Truncate to utilize only the digits individually
                tmp = int(curr_num % 10)
                curr_num = int(curr_num / 10)

        total_sum += (curr_num + tmp)

        # Shift card number over one decimal place
        card_num = card_num / 10

        # When the number is less than one all digits have been processed
        if card_num < 1:
            break

    first_num = int(card_num * 10)
    second_num = int((card_num * 100) % 10)

    print(f"total: {total_sum} % 10 = {total_sum % 10}")

    card_type = "INVALID"

    if round(total_sum % 10) == 0:
        if first_num == 3:
            if card_length == 15 and (second_num == 4 or second_num == 7):
                card_type = "AMEX"
        elif first_num == 4:
            if card_length == 13 or card_length == 16:
                card_type = "VISA"
        elif first_num == 5:
            for i in range(6):
                if card_length == 16 and second_num == i:
                    card_type = "MASTERCARD"

    print(card_type)


main()

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

/*
* Luhn's Algorithm
*
* Multiply every other digit by 2, starting with the number’s second-to-last digit, and then add those products’ digits together.
* Add the sum to the sum of the digits that weren’t multiplied by 2.
* If the total’s last digit is 0 (or, put more formally, if the total modulo 10 is congruent to 0), the number is valid!
*/


int main(void)
{
    long cardNumber = get_long("Number: ");
    double tmpCardNum = (double) cardNumber;
    long totalSum = 0;

    // Counts number of digits.
    int count = 0;
    do
    {
        count++;

        // Set the current number to the last digit available..
        long currentNum = (long) tmpCardNum % 10;

        // On every second digit, multiply them by 2 and separate any digits if needed.
        if (count % 2 == 0)
        {
            currentNum = currentNum * 2;
            int tmpNum = 0;

            // If contains 2 digits, you have to split them to add individually.
            if (currentNum >= 10)
            {
                tmpNum = currentNum % 10;
                currentNum = currentNum / 10;
            }
            totalSum += (currentNum + tmpNum);
        
        else
        {
            // Add the current number to the total sum.
            totalSum += currentNum;
        }

        // Shifting the decimal place over by one,
        // adding one to the count to determine card length.
        tmpCardNum = tmpCardNum / 10;
    }
    while ((int)tmpCardNum != 0);
    // Breaks once all numbers processed : casting to int drops decimal places.


    // Gets the first and second numbers of the card to validate
    // in conjunction with the card length and Luhn's Algorithm.
    int firstNum = (int)(tmpCardNum * 10);
    int secondNum = (int)(tmpCardNum * 100) % 10;
    string cardType = "INVALID";

    if (totalSum % 10 == 0)
    {
        switch (firstNum)
        {
            case 3 :
                // American Express starts with 34 or 37 - 15 digits long
                if (count == 15 && (secondNum == 4 || secondNum == 7))
                {
                    cardType = "AMEX";
                }
                break;
            case 4 :
                // Visa is 13 or 16 digits long.
                if (count == 13 || count == 16)
                {
                    cardType = "VISA";
                }
                break;
            case 5 :
                // MasterCard starts with 51-55 - 16 digits long.
                for (int i = 1; i < 6; i++)
                {
                    if (count == 16 && secondNum == i)
                    {
                        cardType = "MASTERCARD";
                    }
                }
                break;
            default :
                cardType = "INVALID";
        }
    }

    printf("%s\n", cardType);
}



